ons-spark/spark-concepts/partitions.md

```diff
@@ -0,0 +1,941 @@
+
+As the computer scientist Donald Knuth once wrote, 
+
+>Premature optimization is the root of all evil (or at least most of it) in programming
+>
+>    -> *Computer Programming as an Art (1974)*
+
+More importantly it's good practice to be aware of the size of the DataFrame and the number of partitions as this will help to avoid obvious issues or over-partitioning, under-partitioning and highly skewed processes.
+
+A more accurate answer depends on a variety of factors including: the size of the data, data types, distributions within the data, type of processing and other properties defined in the `SparkSession`. Here are some tips for quick wins:
+
+- as suggested above, the first step is getting the code right. Only look to optimise if you're running into performance issues
+- decreasing the `spark.sql.shuffle.partitions` parameter is sensible for smaller datasets
+- if you google it, you might find the optimum size for a Spark partitions is 100-200MB. In practice, this is impossible to keep track of in an analysis script, but it gives you an idea of what is considered big or small
+- it makes sense for the number of partitions to be some multiple of the number of cores in the Spark session. This will help to avoid redundant cores
+
+Remember, it's only worth experimenting on the *optimum* number if you have a real performance issue. 
+
+## Further resources
+
+PySpark:
+
+- [`.rdd.getNumPartitions()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.getNumPartitions.html)
+- [`.rdd.mapPartitions()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.RDD.mapPartitions.html?highlight=mappartitions)
+- [`.coalesce()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.coalesce.html?highlight=coalesce#pyspark.sql.DataFrame.coalesce)
+- [`.repartition()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.repartition.html?highlight=repartition#pyspark.sql.DataFrame.repartition)
+- [`Window.partitionBy()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.Window.partitionBy.html?highlight=partitionby#pyspark.sql.Window.partitionBy)
+- [`DataFrameWriter.partitionBy()`](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrameWriter.partitionBy.html?highlight=partitionby#pyspark.sql.DataFrameWriter.partitionBy)
+
+sparklyr:
+
+- [`sdf_num_partitions()`](https://www.rdocumentation.org/packages/sparklyr/versions/0.8.2/topics/sdf_num_partitions)
+- [`spark_apply()`](https://www.rdocumentation.org/packages/sparklyr/versions/1.7.5/topics/spark_apply)
+- [`sdf_coalesce()`](https://www.rdocumentation.org/packages/sparklyr/versions/1.7.5/topics/sdf_coalesce)
+- [`sdf_repartition()`](https://www.rdocumentation.org/packages/sparklyr/versions/1.7.5/topics/sdf_repartition)
+
+dplyr:
```

Comment:
I've been putting both the dplyr and sparklyr links under the heading sparklyr. Not sure which is better. Perhaps rename to dplyr and sparklyr?

Reply:
good idea